go/pools/fast_pool.go

```diff
@@ -0,0 +1,627 @@
+	}
+
+	isGrowing := capacity > p.state.Capacity
+	p.state.Capacity = capacity
+
+	if isGrowing {
+		p.state.Draining = false
+		return nil
+	}
+
+	p.state.Draining = true
+
+	if block {
+		p.shrink()
+	} else {
+		go p.withLock(p.shrink)
+	}
+
+	return nil
+}
+
+// shrink active resources until capacity it is not above the set capacity.
+// Requires the pool mutex to be locked when calling.
+func (p *FastPool) shrink() {
+	for {
+		remaining := p.active() - p.state.Capacity
+		if remaining <= 0 {
+			p.state.Draining = false
+			if p.state.Capacity == 0 {
+				close(p.pool)
+				p.state.Closed = true
+			}
+			return
+		}
+
+		// We can't remove InUse resources, so only target the pool.
+		// Collect the InUse resources lazily when they're returned.
+		p.Unlock()
+		select {
+		case wrapper := <-p.pool:
+			wrapper.resource.Close()
+			wrapper.resource = nil
+
+			p.withLock(func() {
+				p.state.InPool--
+			})
+
+		case <-time.After(time.Second):
+			// Someone could have pulled from the pool just before
+			// we started waiting. Let's check the pool status again.
+		}
+		p.Lock()
+	}
+}
+
+// SetIdleTimeout sets the idle timeout for resources. The timeout is
+// checked at the 10th of the period of the timeout.
+func (p *FastPool) SetIdleTimeout(idleTimeout time.Duration) {
+	p.withLock(func() {
+		p.state.IdleTimeout = idleTimeout
+	})
+}
+
+// closeIdleResources scans the pool for idle resources
+// and closes them. It is meant to run as a goroutine and will
+// return when the pool is closed.
+func (p *FastPool) closeIdleResources() {
+	for !p.IsClosed() {
+
+		timeout := p.State().IdleTimeout
+
+		if timeout == 0 {
+			// Wait for an updated idleTimeout.
+			time.Sleep(time.Second)
+			continue
+		}
+
+		time.Sleep(timeout / 10)
+
+		var active, minActive int
+		var closed bool
+		for scanning, remaining := true, p.State().InPool; scanning && remaining > 0; remaining-- {
+			p.withLock(func() {
+				active = p.active()
+				minActive = p.state.MinActive
+				closed = p.state.Closed
+			})
+
+			if closed {
+				return
+			}
+
+			if active <= minActive {
+				break
+			}
+
+			select {
+			case wrapper, ok := <-p.pool:
+				if !ok {
+					return
+				}
+
+				deadline := wrapper.timeUsed.Add(timeout)
+				if time.Now().After(deadline) {
+					p.withLock(func() {
+						p.state.IdleClosed++
+						p.state.InPool--
+					})
+
+					wrapper.resource.Close()
+					wrapper.resource = nil
+					break
+				}
+
+				// Not expired--back into the pool we go.
+				select {
+				case p.pool <- wrapper:
+				default:
+					// Can't put back into pool. Might be full.
+					p.withLock(func() {
+						p.state.InPool--
+					})
+
+					wrapper.resource.Close()
+					wrapper.resource = nil
+
+					scanning = false
+				}
+
+			default:
+				// The pool might have been used while we were iterating.
+				// Maybe next time!
+				scanning = false
+			}
+		}
+	}
+}
+
+func (p *FastPool) withLock(f func()) {
+	p.Lock()
+	f()
+	p.Unlock()
```

Comment:
This needs to be in a defer. Otherwise a panic in `f` will deadlock the pool.

Reply:
👍